### PR TITLE
fs/smartfs : Add block write function for journaling

### PR DESCRIPTION
**Pull request by Taejun-Kwon**
Currently, journaling uses bytewrite operation.
But not all architecture support bytewrite, so add this.

### CODE REVIEW

File: os/fs/driver/mtd/smart.c [hunk 1]
```diff
@@ -5635,7 +5663,11 @@ static int smart_journal_recovery(FAR struct smart_struct_s *dev, journal_log_t
...
 			/* If crc is same, then update mtd header */
 			fvdbg("valid data, update header here..\n");
+#ifdef CONFIG_MTD_BYTE_WRITE
 			ret = MTD_WRITE(dev->mtd, psector * dev->sectorsize, mtd_size, (FAR uint8_t *)&log->mtd_header);
+#else
+			ret = smart_byte_to_block_write(dev, psector * dev->sectorsize, mtd_size, (FAR uint8_t *)&log->mtd_header);
+#endif
 			if (ret != mtd_size) {
 				fdbg("Checkout error, write mtd header.. ret : %d\n", ret);
 				return -EIO;
```
Comment (amogh-samsung) [inline os/fs/driver/mtd/smart.c:5673]: Looks Good

Comment (amogh-samsung) [inline os/fs/driver/mtd/smart.c:5673]: Looks Good

**Review by amogh-samsung — APPROVED** [review]

**Review by jeongchanKim — APPROVED** [review]
LGTM

File: os/fs/driver/mtd/smart.c [hunk 2]
```diff
@@ -5162,6 +5180,11 @@ static int smart_journal_process_transaction(FAR struct smart_struct_s *dev, jou
 
 	case SMART_JOURNAL_TYPE_COMMIT: {
 
+#ifdef CONFIG_MTD_BYTE_WRITE
+		/** If board supports bytewrite, bytewrite always faster than block write.
```
Comment (thapav) [inline os/fs/driver/mtd/smart.c:5184]: bytewrite is* always

File: os/fs/driver/mtd/smart.c [hunk 3]
```diff
@@ -1341,6 +1341,41 @@ static int smart_setsectorsize(FAR struct smart_struct_s *dev, uint16_t size)
...
+ *
+ ****************************************************************************/
+static int smart_byte_to_block_write(FAR struct smart_struct_s *dev, size_t offset, int nbytes, FAR const uint8_t *buffer)
+{
+	uint32_t startblock;
+	uint16_t nblocks;
+	int ret = OK;
+	/* First calculate the start block and number of blocks affected. */
+	
+	startblock = offset / dev->geo.blocksize;
```
Comment (davidfather) [inline os/fs/driver/mtd/smart.c:1359]: Is there any possibility that dev->geo.blocksize would be zero?

**Comment by Taejun-Kwon** [comment]
Powercut test works properly during 3days.

File: os/fs/driver/mtd/smart.c (hunk 3 above)
Comment (Taejun-Kwon) [reply os/fs/driver/mtd/smart.c:1359]: yes, it will never happen.

File: os/fs/driver/mtd/smart.c (hunk 2 above)
Comment (Taejun-Kwon) [reply os/fs/driver/mtd/smart.c:5184]: Thanks.

**Review by sunghan-chang — APPROVED** [review]

**Pull request merged**